### PR TITLE
[reference only -- do not merge] adding scaling and kweighting utilities

### PR DESCRIPTION
**Pull request by alisiafadini**
Added `kweighting.py` and `map_utils.py` for difference map calculation, scaling and k-weighted map generation.

Kweight optimization is based on the TV regularization optimization in `tv.py`

Still needs unit testing for the kweighting.

### CODE REVIEW

File: meteor/settings.py [hunk 1]
```diff
@@ -2,3 +2,4 @@
 TV_STOP_TOLERANCE: float = 0.00000005
 TV_MAX_NUM_ITER: int = 50
 TV_MAP_SAMPLING: int = 3
+KWEIGHTED_MAP_SAMPLING: int = 4
```
Comment (tjlane) [inline meteor/settings.py:5]: hmmm why? i guess we need to figure out all this sampling stuff

File: meteor/map_utils.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    reference (rs.DataSeries): Single-column DataSeries to use as the reference for scaling.
+    dataset_to_scale (rs.DataSeries): Single-column DataSeries to be scaled.
+    inplace (bool): If `True`, modifies the original DataSeries. If `False`,
+    returns a new scaled DataSeries.
+
+    Returns:
+    None if `inplace` is True, otherwise rs.DataSeries with scaled data.
+    """
+
+    def aniso_scale_func(params, x_ref, x_scale, miller_indices):
```
Comment (tjlane) [inline meteor/map_utils.py:53]: use better names than `x_ref` and `x_scale`, and type hint this function

Comment (tjlane) [reply meteor/map_utils.py:53]: also it would be best to include `objective` or similar in the function name

File: meteor/map_utils.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    hk_prod, hl_prod, kl_prod = h * k, h * l, k * l
+
+    t = -(
+        h_sq * result.x[1]
+        + k_sq * result.x[2]
+        + l_sq * result.x[3]
+        + 2 * hk_prod * result.x[4]
+        + 2 * hl_prod * result.x[5]
+        + 2 * kl_prod * result.x[6]
+    )
```
Comment (tjlane) [inline meteor/map_utils.py:101]: also I hope `ruff` yells at you for having a mysterious variable named `t` ;)

File: meteor/map_utils.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        )
+
+        return x_ref - params[0] * np.exp(t) * x_scale
+
+    reference_data = reference.to_numpy()
+    scale_data = dataset_to_scale.to_numpy()
+    uncertainties_data = uncertainties.to_numpy() if uncertainties is not None else None
+
+    miller_indices_ref = np.array(list(reference.index))
+    miller_indices_scale = np.array(list(dataset_to_scale.index))
```
Comment (tjlane) [inline meteor/map_utils.py:74]: do you really need to cast to `list` here first? if so add a comment

File: meteor/map_utils.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        h_sq * result.x[1]
+        + k_sq * result.x[2]
+        + l_sq * result.x[3]
+        + 2 * hk_prod * result.x[4]
+        + 2 * hl_prod * result.x[5]
+        + 2 * kl_prod * result.x[6]
+    )
+
+    scale_factor = result.x[0] * np.exp(t)
+    scaled_data = scale_factor * scale_data
```
Comment (tjlane) [inline meteor/map_utils.py:104]: This block is repeated and hard to read -- pull it out into a separate function that's clearly written.

File: meteor/map_utils.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+            uncertainties[:] = scaled_uncertainties
+        return None
+    else:
+        scaled_dataset = dataset_to_scale.copy()
+        scaled_dataset[:] = scaled_data
+        if uncertainties is not None:
+            scaled_uncertainties_series = uncertainties.copy()
+            scaled_uncertainties_series[:] = scaled_uncertainties
+            return scaled_dataset, scaled_uncertainties_series
+        return scaled_dataset
```
Comment (tjlane) [inline meteor/map_utils.py:123]: These nested if/else blocks are hard to read -- can you make it cleaner?

File: meteor/map_utils.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+def compute_amplitude_fofo_difference(
+    data1: rs.DataSeries,
+    data2: rs.DataSeries,
+    data3: rs.DataSeries,
+    uncertainties1: Optional[rs.DataSeries] = None,
+    uncertainties2: Optional[rs.DataSeries] = None,
+) -> Union[rs.DataSeries, tuple[rs.DataSeries, rs.DataSeries]]:
```
Comment (tjlane) [inline meteor/map_utils.py:131]: `Optional` and `Union` are old now, use `|` as an "or" operator, where `| None` gives you a way to declare something is optional.

File: meteor/map_utils.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        if uncertainties is not None:
+            scaled_uncertainties_series = uncertainties.copy()
+            scaled_uncertainties_series[:] = scaled_uncertainties
+            return scaled_dataset, scaled_uncertainties_series
+        return scaled_dataset
+
+def compute_amplitude_fofo_difference(
+    data1: rs.DataSeries,
+    data2: rs.DataSeries,
+    data3: rs.DataSeries,
```
Comment (tjlane) [inline meteor/map_utils.py:128]: so sad. what are all the datas. why not `data6`.

File: meteor/map_utils.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    uncertainties2: Optional[rs.DataSeries] = None,
+) -> Union[rs.DataSeries, tuple[rs.DataSeries, rs.DataSeries]]:
+    """
+    First, scale data1 and data2 to the common scale defined by data3,
+    then compute the difference (data2 - data1). Optionally, propagate uncertainties.
+
+    Parameters:
+    data1 (rs.DataSeries): First dataset to be used in difference calculation (e.g., F_off).
+    data2 (rs.DataSeries): Second dataset to be used in difference calculation (e.g., F_on).
+    data3 (rs.DataSeries): Reference dataset used for scaling data1 and data2 (e.g., F_calc).
```
Comment (tjlane) [inline meteor/map_utils.py:139]: why can't we just scale two datasets to one another?? I don't think we need F_calc do we. Marius does this craziness but I never understood why.

File: meteor/map_utils.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    Returns:
+    rs.DataSeries or tuple: The difference (data2 - data1) after scaling to the reference scale.
+    If uncertainties are provided, returns a tuple (difference, propagated_uncertainties).
+    """
+
+    # Ensure that both or neither uncertainties are provided
+    if (uncertainties1 is not None) != (uncertainties2 is not None):
+        raise ValueError("Either provide uncertainties for both data1 and data2, or provide none.")
+
+    if uncertainties1 is not None and uncertainties2 is not None:
```
Comment (tjlane) [inline meteor/map_utils.py:152]: given all the hoops you have to jump through to handle the with and without uncertainties thing, I would just make two different functions -- one taking the uncertainties and one not

File: meteor/map_utils.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+        # Compute the difference between the scaled data2 and data1
+        difference = scaled_data2 - scaled_data1
+        return difference
+
+
+def compute_fofo_difference_map(
+    data1: rs.DataSeries,
+    data2: rs.DataSeries,
+    f_calcs: rs.DataSeries,
+    phases: rs.DataSeries,
```
Comment (tjlane) [inline meteor/map_utils.py:182]: see all comments above about these arguments...

File: test/unit/test_map_utils.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
+import numpy as np
+import pandas as pd
+import pytest
+import reciprocalspaceship as rs
+
+from meteor import map_utils
+
+
+def generate_mock_dataset(miller_indices, data):
```
Comment (tjlane) [inline test/unit/test_map_utils.py:9]: I'll review the tests separately in a 2nd step

File: meteor/kweighting.py [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+import reciprocalspaceship as rs
+from scipy.optimize import minimize_scalar
+
+from .map_utils import compute_amplitude_fofo_difference
+from .settings import KWEIGHTED_MAP_SAMPLING
+from .utils import compute_map_from_coefficients
+from .validate import negentropy
+
+
+@dataclass
```
[the remark this answers is not in the document]
Comment (alisiafadini) [reply meteor/kweighting.py:14]: do you mean a common class for kweighting, TV, other?

File: meteor/kweighting.py [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    optimal_negentropy: float
+
+def compute_weights(df: rs.DataSeries, sigdf: rs.DataSeries, kweight: float) -> rs.DataSeries:
+    """
+    Compute weights for each structure factor based on DeltaF and its uncertainty.
+    """
+    w = 1 + (sigdf**2 / (sigdf**2).mean()) + kweight * (df**2 / (df**2).mean())
+    return w**-1
+
+def compute_kweighted_difference_map(
```
[the remark this answers is not in the document]
Comment (alisiafadini) [reply meteor/kweighting.py:26]: That's fine. It just means you have to provide labels for the right column every time. Do we care about the number of inputs to a function? In my mind I was hoping to minimize this

File: meteor/kweighting.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+    data: tuple[rs.DataSeries, rs.DataSeries],
+    uncertainties: tuple[rs.DataSeries, rs.DataSeries],
+    f_calcs: rs.DataSeries,
+    phases: rs.DataSeries,
+    kweight: float,
+    crystal_params: dict,
```
[the remark this answers is not in the document]
Comment (alisiafadini) [reply meteor/kweighting.py:32]: > First: can't I use any weight? Doesn't need to be a k-weight, right?
it's within a k-weighting function right?

> For crystal_params, a structured input would be much better. Actually just requiring that the rs.DataSet has a valid cell attribute makes things totally unambiguous and robust.
I agree actually. This is a good idea

File: meteor/kweighting.py [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+    uncertainties: tuple[rs.DataSeries, rs.DataSeries],
+    f_calcs: rs.DataSeries,
+    phases: rs.DataSeries,
+    crystal_params: dict,
+    full_output: Literal[True],
+    kweight_values_to_scan: Sequence[float] | None = None,
+) -> tuple[rs.DataSet, KWeightingOptimizationResult]:
+    ...
+
+def optimize_kweight_for_kweighting(
```
[the remark this answers is not in the document]
Comment (alisiafadini) [reply meteor/kweighting.py:111]: Let's work on it together!

File: meteor/map_utils.py (hunk 8 above)
Comment (alisiafadini) [reply meteor/map_utils.py:128]: that's fair that's fair. my bad

File: meteor/map_utils.py (hunk 9 above)
Comment (alisiafadini) [reply meteor/map_utils.py:139]: image

I admit I inherited this from Marius, but intuitively, I assumed that  a common reference dataset would help in normalizing  relative to a stable baseline if the two datasets are noisy.

I ran a small test with two datasets at different noise levels

```
x = np.linspace(0, 10, 100)  # Common x-axis
y1 = 2 * x + 5 + np.random.normal(0, 1, 100)  
y2 = 2 * x + 5 + np.random.normal(0, 5, 100)
```

and scaled using 

```
def scale_to_each_other(y1, y2):
    coef = np.polyfit(y1, y2, 1)  # Fit y2 as a function of y1
    y2_scaled = coef[0] * y1 + coef[1]  # Scale y2 to y1
    return y2_scaled, coef
```
first to each other, then to the perfect `reference = 2 * x + 5 `. Seems like it is indeed better to scale to the common reference?

File: meteor/settings.py (hunk 1 above)
Comment (alisiafadini) [reply meteor/settings.py:5]: This actually is a slip. I think we should pick one sampling and stick with it... tradeoff between compute and quality I guess?

**Pull request closed without merging**